crates/vm2-interface/src/state_interface.rs

```diff
@@ -109,33 +164,49 @@ impl HeapId {
         Self(value)
     }
 
+    /// Converts this ID to an integer value.
     pub const fn as_u32(self) -> u32 {
         self.0
     }
 }
 
+/// Event emitted by EraVM.
+///
 /// There is no address field because nobody is interested in events that don't come
-/// from the event writer, so we simply do not record events coming frome anywhere else.
-#[derive(Clone, PartialEq, Debug)]
+/// from the event writer, so we simply do not record events coming from anywhere else.
+#[derive(Debug, Clone, Copy, PartialEq)]
 pub struct Event {
+    /// Event key.
     pub key: U256,
+    /// Event value.
     pub value: U256,
+    /// Is this event first in a chain of events?
     pub is_first: bool,
+    /// Shard identifier (currently, always set to 0).
     pub shard_id: u8,
+    /// 0-based index of a transaction that has emitted this event.
     pub tx_number: u16,
 }
 
-#[derive(Debug)]
+/// L2-to-L1 log emitted by EraVM.
+#[derive(Debug, Clone, Copy, PartialEq)]
 pub struct L2ToL1Log {
+    /// Log key.
     pub key: U256,
+    /// Log value.
     pub value: U256,
+    /// Is this a service log?
```

Comment:
I actually don't understand where the is_service terminology comes from. At least with events it is sometimes used interchangeably with is_first.